crates/subgraph/src/apy.rs

```diff
@@ -0,0 +1,777 @@
+            .filter(|v| {
+                u64::from_str(&v.timestamp.0).unwrap()
+                    <= u64::from_str(&first_trade.timestamp.0).unwrap() + DAY
+            })
+            .collect::<Vec<&&Trade>>()[0];
+
+        // vaults starting capital at end of first day of its first ever trade
+        let starting_capital = if first_day_last_trade
+            .input_vault_balance_change
+            .vault
+            .vault_id
+            .0
+            == vol.id
+            && first_day_last_trade.input_vault_balance_change.vault.token == vol.token
+        {
+            U256::from_str(
+                &first_day_last_trade
+                    .input_vault_balance_change
+                    .new_vault_balance
+                    .0,
+            )?
+        } else {
+            U256::from_str(
+                &first_day_last_trade
+                    .output_vault_balance_change
+                    .new_vault_balance
+                    .0,
+            )?
+        };
+
+        // the time range for this token vault
+        let mut start = u64::from_str(&first_trade.timestamp.0)?;
+        start_timestamp.inspect(|t| {
+            if start > *t {
+                start = *t;
+            }
+        });
+        let end = end_timestamp.unwrap_or(chrono::Utc::now().timestamp() as u64);
+
+        // this token vault apy
+        let apy = if starting_capital.is_zero() {
+            0_f64
+        } else {
+            // by 4 point decimals
+            let change_ratio = i64::try_from(
+                vol.net_vol
+                    .saturating_mul(I256::from_raw(U256::from(10_000)))
+                    .checked_div(I256::from_raw(starting_capital))
+                    .unwrap_or(I256::ZERO),
+            )? as f64;
+            let time_to_year_ratio = ((end - start) as f64) / YEAR as f64;
+            (change_ratio * time_to_year_ratio) / 100f64
+        };
+        token_vaults_apy.push(TokenVaultAPY {
+            id: vol.id.clone(),
+            token: vol.token.clone(),
+            start_time: start,
+            end_time: end,
+            net_vol: vol.net_vol,
+            apy,
+            capital: starting_capital,
+        });
+    }
+
+    Ok(token_vaults_apy)
+}
+
+/// Calculates an order's pairs' ratios from their last trades in a given list of trades
+fn get_pairs_ratio(order_apy: &OrderAPY, trades: &[Trade]) -> HashMap<TokenPair, Option<I256>> {
+    let one = I256::from_str(ONE).unwrap();
+    let mut pair_ratio_map: HashMap<TokenPair, Option<I256>> = HashMap::new();
+    for input in &order_apy.inputs_token_vault_apy {
+        for output in &order_apy.outputs_token_vault_apy {
+            if input.token != output.token {
+                // find this pairs trades from list of order's trades
+                let pair_trades = trades
+                    .iter()
+                    .filter(|v| {
+                        v.input_vault_balance_change.vault.token == input.token
```

Comment:
does this handle the case where the most recent trade for the pair is the reverse direction? so the same tokens but flipped i/o?

in that case you invert the price to use it

Reply:
good catch, I'll add